### ISSUE TITLE
Scheduling Hook LoRA incorrect CLIP cache in subsequent run after value change

### ISSUE THREAD

**Issue by tajino** [comfyanonymous/ComfyUI#7195]
### Expected Behavior

Scheduling Hook LoRA and modification to value in subsequent run should produce output that is reproducable as when loaded fresh without cache.

output 2 should be identical to output 3 as there are no change to the values, except freeing the cache and run fresh.

![Image](https://github.com/user-attachments/assets/22a63f75-989e-4695-bbe6-cd649ab38fba)

### Actual Behavior

With Scheduling Hook LoRA, any value change to the Hook Lora nodes will produce incorrect unreproducable output on subsequent cached run compared to right after freeing model and node cache.

### Steps to Reproduce

Load the attached workflow, steps are annotated in the workflow notes.
[Hook Lora cached Clip issue.json](https://github.com/user-attachments/files/19197827/Hook.Lora.cached.Clip.issue.json)

or drop this image into ComfyUI

![Image](https://github.com/user-attachments/assets/29501d78-8c4a-4ce1-81a8-9eff79c32b88)

This workflow is specifically created to exhibit this issue in the simplest form.

Load this lora https://civitai.com/models/180780?modelVersionId=379351
or any other SDXL LoRA that you have into both Load LoRA and Create Hook LoRA, can leave both strength model and strength clip at 1

1. Run the workflow as a baseline
2. After the first Run, change start_percent of this Create Hook Keyframe to 0.7 or any other values and run the workflow again.
3. When done, click ComfyUI Manager "free model and node cache" (the solid vacuum cleaner icon)
4. Don't change any other values, run the workflow again.
5. Check ComfyUI\temp\ folder and compare the last 2 output, they should be identical, but they are not.

It looks like something went wrong with the cached CLIP.

### Debug Logs

```powershell
unrelated
```

### Other

This issue only happens if the CLIP that goes into Set CLIP Hooks does not come straight from a Checkpoint loader.

This cache issue should also affect all Hook LoRA related nodes.

**Comment by comfyanonymous** [comment]
@Kosinkadink

**Comment by Kosinkadink** [comment]
I tried to reproduce your issue across both the current version of ComfyUI and prior to a recent PR of mine that related to hooks, and I cannot get the issue you run into. However, one difference between your results and mine is that I never get anything resembling output 2 - changing the value of the keyframe to 0.7 immediately gives me something very close to output 3. Using manager's model unload functions then gives me back output 3 again.

![Image](https://github.com/user-attachments/assets/fa59549f-0a72-4b79-b682-8e3c576816dd)

On your end, assuming you've already updated ComfyUI to the latest version, can you set your values to your output 2, and run it right after a ComfyUI restart directly rather than after changing the values after output 1?

**Comment by tajino** [comment]
Thanks for looking into this. 

> On your end, assuming you've already updated ComfyUI to the latest version.

I'm using ComfyUI 0.3.26 on an RTX 3070 8GB VRAM, 128GB system RAM, AMD TR 3970X CPU. I've noticed these issue in the previous ComfyUI v0.3.12-1-g7fc3ccdc (2025-01-16) about the time when Hook Lora was newly introduced.

> can you set your values to your output 2, and run it right after a ComfyUI restart directly rather than after changing the values after output 1?

Setting to Output 2 value (0.7) and running it right after a ComfyUI restart will always get Output 3, same as clearing the cache will also get Output 3, only the subsequent run with a new value and with cache still on, it will give an output that is not reproducable after a restart or clear cache.

Not sure if this is due to offloading and cache. Also there is a warning message in my log in the initial run after a fresh ComfyUI restart/start, about "1Torch was not compiled with flash attention", if that is related.

> changing the value of the keyframe to 0.7 immediately gives me something very close to output 3.

When you said "very close", does that mean it is not exactly identical? and only after clearing cache you get the output 3 exactly identical to my output 3?

here is my console log:

![Image](https://github.com/user-attachments/assets/b4447e2e-e56d-421a-9a7c-c33a4e4a5b2d)


```
got prompt
model weight dtype torch.float16, manual cast: None
model_type EPS
Using pytorch attention in VAE
Using pytorch attention in VAE
VAE load device: cuda:0, offload device: cpu, dtype: torch.bfloat16
CLIP/text encoder model load device: cuda:0, offload device: cpu, current: cpu, dtype: torch.float16
Requested to load SDXLClipModel
loaded completely 5680.8 1560.802734375 True
D:\_SD_ComfyUI\ComfyUI_windows_portable\ComfyUI\comfy\ldm\modules\attention.py:441: UserWarning: 1Torch was not compiled with flash attention. (Triggered internally at ..\aten\src\ATen\native\transformers\cuda\sdp_utils.cpp:455.)
  out = torch.nn.functional.scaled_dot_product_attention(q, k, v, attn_mask=mask, dropout_p=0.0, is_causal=False)
Requested to load SDXL
loaded completely 5395.290146636963 4897.0483474731445 True
100%|████████████████████████████████████████████████████| 20/20 [00:12<00:00,  1.61it/s]
Requested to load AutoencoderKL
loaded completely 1045.445167541504 159.55708122253418 True
Prompt executed in 22.56 seconds
got prompt
Requested to load SDXLClipModel
loaded completely 3325.68808631897 1560.802734375 True
loaded completely 5181.633812713623 4897.0483474731445 True
100%|████████████████████████████████████████████████████| 20/20 [00:11<00:00,  1.79it/s]
Requested to load AutoencoderKL
loaded completely 1049.947608947754 159.55708122253418 True
Prompt executed in 17.04 seconds
got prompt
model weight dtype torch.float16, manual cast: None
model_type EPS
Using pytorch attention in VAE
Using pytorch attention in VAE
VAE load device: cuda:0, offload device: cpu, dtype: torch.bfloat16
CLIP/text encoder model load device: cuda:0, offload device: cpu, current: cpu, dtype: torch.float16
Requested to load SDXLClipModel
loaded completely 5574.79991607666 1560.802734375 True
Requested to load SDXL
loaded completely 5454.174828338623 4897.0483474731445 True
100%|████████████████████████████████████████████████████| 20/20 [00:11<00:00,  1.76it/s]
Requested to load AutoencoderKL
loaded completely 812.4012222290039 159.55708122253418 True
Prompt executed in 20.81 seconds
```

**Comment by tajino** [comment]
I have done more tests, tried with the newer versions of pytorch 2.6.0+cu124 and 2.6.0+cu126. The previous test was done with pytorch 2.3.0+cu121, every pytorch version produces different results, but consistently the second run without freeing cache or restarting ComfyUI produced a 'wrong' result.

you can see in the comparison between pytorch versions, doing the same, first run at 0.3, change to 0.7 run it and clear cache and run reproduction steps.

Strangely, output 2 of the one that ComfyUI ran with --disable-all-custom-nodes, is giving a different result than with custom nodes. Both are using the exact same pytorch 2.6.0+cu124. How would custom nodes be even able to affect the result when the workflow is not using any custom nodes?

![Image](https://github.com/user-attachments/assets/35cb7fc4-bdf0-4491-9783-e2650e4e5517)

Running ComfyUI with --disable-all-custom-nodes, the console log has got this error at the first run :

![Image](https://github.com/user-attachments/assets/b1af0427-0d0d-4d84-8cb7-2b5b8c4ea1d8)

```indows-standalone-build --listen --disable-auto-launch --disable-all-custom-nodes
Checkpoint files will always be loaded safely.
Total VRAM 8192 MB, total RAM 130945 MB
pytorch version: 2.6.0+cu124
Set vram state to: NORMAL_VRAM
Device: cuda:0 NVIDIA GeForce RTX 3070 : cudaMallocAsync
Using pytorch attention
ComfyUI version: 0.3.26
ComfyUI frontend version: 1.11.8
[Prompt Server] web root: D:\_SD_ComfyUI\ComfyUI_windows_portable\python_embeded\Lib\site-packages\comfyui_frontend_package\static
D:\_SD_ComfyUI\ComfyUI_windows_portable\python_embeded\Lib\site-packages\kornia\feature\lightglue.py:44: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  @torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
Skipping loading of custom nodes
Starting server

To see the GUI go to: http://0.0.0.0:8188
To see the GUI go to: http://[::]:8188
got prompt
model weight dtype torch.float16, manual cast: None
model_type EPS
Using pytorch attention in VAE
Using pytorch attention in VAE
VAE load device: cuda:0, offload device: cpu, dtype: torch.bfloat16
CLIP/text encoder model load device: cuda:0, offload device: cpu, current: cpu, dtype: torch.float16
Requested to load SDXLClipModel
loaded completely 5680.8 1560.802734375 True
Requested to load SDXL
loaded completely 5394.145615386963 4897.0483474731445 True
  0%|                                                             | 0/20 [00:00<?, ?it/s]D:\_SD_ComfyUI\ComfyUI_windows_portable\python_embeded\Lib\site-packages\torchsde\_brownian\brownian_interval.py:608: UserWarning: Should have tb<=t1 but got tb=14.614643096923828 and t1=14.614643.
  warnings.warn(f"Should have {tb_name}<=t1 but got {tb_name}={tb} and t1={self._end}.")
100%|████████████████████████████████████████████████████| 20/20 [00:10<00:00,  1.84it/s]
Requested to load AutoencoderKL
loaded completely 1203.5585289001465 159.55708122253418 True
Prompt executed in 19.84 seconds
got prompt
Requested to load SDXLClipModel
loaded completely 3483.8014514923098 1560.802734375 True
loaded completely 5190.489365386963 4897.0483474731445 True
100%|████████████████████████████████████████████████████| 20/20 [00:10<00:00,  1.84it/s]
Requested to load AutoencoderKL
loaded completely 1195.5585289001465 159.55708122253418 True
Prompt executed in 16.36 seconds
```

I did further tests, I alternate the Create Hook Keyframe node value between 0.3 and 0.7 on each subsequent run without clearing cache or restarting ComfyUI, here is what I got, the output got wilder and wilder, looks like the cached weights is accumulated on each run.

![Image](https://github.com/user-attachments/assets/ee27d6ab-ddd9-4fbe-a80f-1a826e39961e)

Would be great if you could take a look at this again. 

What pytorch version+cu version are you using in the previous test that got indentical output 2 and output 3 result? Just to confirm, you did run the output 2 right after output1 completed and change value to 0.7 and run it straight away without clearing cache or restarting ComfyUI right?

**Comment by tajino** [comment]
Here are some of the dependencies versions :

torch : 2.6.0+cu124
torchvision : 0.21.0+cu124
torchaudio : 2.6.0+cu124
torchsde : 0.2.6
numpy : 1.26.4
einops : 0.8.0

**Comment by tajino** [comment]
@Kosinkadink 

Another test of alternating the values between each subsequent run, with ComfyUI --disable-all-custom-nodes.

Output 1,3,5,7 should be identical, and Output 2,4,6 should be identical because they are of exact same values, but they are not.

Which python packages that are most likely related or responsible for this incorrect cache accumulation?

![Image](https://github.com/user-attachments/assets/ed391b9f-8f4d-4df3-8828-cba48e0d23fc)

**Comment by tajino** [comment]
Another test, change prompt 'purple' to 'green' and alternating it, subsequent runs without clearing cache. Something is definitely going on with the cache.

![Image](https://github.com/user-attachments/assets/eb8a57d5-e806-4f2d-b10f-5d7264233adb)

Same seed, same value, different result each time, getting more wild with subsequent runs. 

If I bypass the 'Set CLIP Hooks' node, and do the same, alternating between 'purple' and 'green', I get correct repeatable outputs, see below.

Also, I need to mention that if you alternate the seed between 10 and 11, without changing any values of Hook Lora nodes or prompt, the output is correct and repeatable. Anything that changes the conditioning will result in wrong output over subsequent runs.

![Image](https://github.com/user-attachments/assets/369f7dbe-a900-48dd-bf39-1bcec4c31b63)

![Image](https://github.com/user-attachments/assets/fb76e24d-076e-4022-9868-4645dd4eb152)

**Comment by Kosinkadink** [comment]
TL;DR I am able to reproduce your issue, but it is specific to relatively low vram GPUs (8GB, tested using RTX 3070 and Mobile RTX 4070) AND specifically SDXL clip - issue does not occur on SD1.5. Higher VRAM devices do not experience the bug, despite the logs for both 8GB and 24GB (RTX 4090) both stating models were loaded completely. Tricking the code to make ComfyUI think the RTX 4090 has only 8GB of VRAM does not reproduce the problem either. Reproduction also specifically requires that a lora is applied in the 'normal' way alongside the hooks - using just hooks for the LoRA weights has no issues.

Reproduction of the bug requires these 4 things; if you don't have one of these, you will not have the bug:
1) A GPU with ~8GB of VRAM
2) Using an SDXL model (specifically, SDXL clip, but likely any multi-CLIP architecture)
3) Applying LoRAs both the static way (Load LoRA nodes) and dynamic way (Create Hook LoRA) simultaneously 
4) Changing ANY value related to Set CLIP Hooks after one run while keeping the model loaded - it doesn't even have to be the keyframes passed into Set Hook Keyframes, you could toggle ```apply_to_conds``` instead

Everything else is a red herring. The weights are applied properly to the base model, but not SDXL CLIP after any value changes in the above scenario. I confirmed this by using ```Cond Pair Set Props``` node to set the hooks w/ a schedule only on the conditioning that the model will see rather than the CLIP, or by setting ```clip_strength``` to 0.0 on the Load LoRA node. Doing something unrelated to the CLIP weights, like using ```CLIP Set Last Layer``` node and changing the value between runs does not reproduce the issue, so it is specifically about the weight patches. My hunch is the fact that SDXL uses a dual clip, something goes wrong with unloading the patches, buy I have no idea why it happens specifically on lower VRAM devices despite them having ample VRAM to load weights.

I happened to have a test PC with an RTX 3070, so luckily I was able to do a direct apples-to-apples testing attempt.

To clarify what I said in my original post, I did not have the bug at all - the 4090 did not have the conditions necessary to trigger the bug, so my output 2 and output 3 were pixel identical. My actual results for output 1/output 3 looked slightly different than yours likely due to difference in pytorch version, but the consistency of the results was nominal.

I'll try to make a fix for this today if I can figure out why the weights are not cleaned up for multi-CLIP models on 8GB devices.

**Comment by Kosinkadink** [comment]
Update: I figured out the root cause - ```ModelPatcher.partially_unload``` does not account for hook patches, so when hooks and weight patches are both present at once, it screws up the references to the proper weights so they don't get unpatched all the way to their original status. Even though the logs say the model gets loaded completely, models can be requested to be unloaded partially when another is loaded without a log statement being printed.

Not sure how long it will take for me to make this work, but hopefully today as originally planned.

**Comment by Kosinkadink** [comment]
I submitted a PR to fix the bug you found, feel free to test that branch to confirm it fixes the issue on your end.

**Comment by tajino** [comment]
Tested the PR, it is working correctly, thanks a lot for fixing this, really appreaciate it.

It is very lucky that you still have a RTX 3070 8GB, otherwise it would have been an unreproducible mystery, that will eventually get solved with a hardware change to a bigger VRAM card unknowingly.
